runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -812,21 +796,35 @@ public boolean isForWindow(
 
     private final List<String> userStateNames;
     private final Coder windowCoder;
-    private final StateInternals stateInternals;
+    private final ArrayDeque<KV<ByteBuffer, BoundedWindow>> cleanupQueue;
+    private final Supplier<ByteBuffer> keyedStateBackend;
 
-    StateCleaner(List<String> userStateNames, Coder windowCoder, StateInternals stateInternals) {
+    StateCleaner(
+        List<String> userStateNames, Coder windowCoder, Supplier<ByteBuffer> keyedStateBackend) {
       this.userStateNames = userStateNames;
       this.windowCoder = windowCoder;
-      this.stateInternals = stateInternals;
+      this.keyedStateBackend = keyedStateBackend;
+      this.cleanupQueue = new ArrayDeque<>();
```

Comment:
I think we have a consistency problem here because this is not persisted in checkpoints.

Reply:
Hmm, that should not be the case when the cleanup is performed as part of finishing the bundle. Need to see how that is possible.

Reply:
Right, if after `invokeFinishBundle()` has been called the cleanup queue is always cleared as part of however many additional bundles, then we are fine. If we delay any cleanup for after the checkpoint has completed, then we would have a problem.